README.md
KoranFinder
============

KoranFinder is a fork of
[QuranFinder](https://github.com/SafaAlfulaij/QuranFinder) which is a Limnoria
plugin to get Quran verses using Global Quran API.

## History
It was first time developed by a ##islam channel member on freenode. a few days
later it was
moved to Github. Now that I check the first commit was Mar 27, 2016. After about 2 years I forked it and changed its name to `KoranFinder` so
I can maintain it myself(it seems they don't have much time maintaining this).

## Usage

```
Mybot: quran <surah> <ayah/ayat> [lang]
```
returns ayah number \<ayah\> of surah number \<surah\> in \<lang\> language.
By default, `lang` is `en` which refers to English.
To get just one ayah you should go like this:

```
Mybot: quran 1 1 en
```
Which returns 
```
1,1: Praise be to Allah, Lord of the Worlds,
```
You could also get more than one ayah, like:
```
Mybot: quran 1 1-3
Mybot: quran 1 1,2,4
```
Which returns ayat 1 to 3 from surah 1 and ayat 1,2,4 from surah 1.
You can also combine them:
```
Mybot: quran 1 1-3,5
```
Which returns ayat 1 to 3 and also 5 from surah 1.

The bot can obtain translation of the Quran for different languages. (ar, en, fa, tr) are the main ones. You can get extra translations and other data sources by using the data source key provided by Global Quran API. For more information see the [Wiki](https://github.com/SafaAlfulaij/QuranFinder/wiki)

For example
```
Mybot: quran 1 1 en.hilali
```
provides the Hilali English translation of the Quran.

## Licence
This program is free software under GNU General Public License 3 or any later version and COMES WITHOUT ANY WARRANTY, for more information see LICENCE.

## To-Do
- [ ] Don't send a message for each ayah and split messages, instead use supybot's more(works for English but not Arabic)
- [ ] Use config.py for the maximum difference between two verses which by default is 5.


CHANGELOG.md
# ChangeLog

Want to know before October 12? see [QuranFinder ChangeLog](https://github.com/SafaAlfulaij/QuranFinder/blob/master/CHANGELOG.md)
 - October 12, 2017: KoranFinder was born!
 - March 28, 2018: the part of KoranFinder which gets verses from GlobalQuran
   is now a seperate module named [PyGQ](https://github.com/farooqkz/PyGQ).
 - May 27, 2018: KoranFinder uses new version of PyGQ


test.py
'''
    Copyright 2017-2018 Farooq Karimi Zadeh <farooghkarimizadeh@gmail.com>

    This file is part of KoranFinder.

    KoranFinder is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this software.  If not, see <http://www.gnu.org/licenses/>.

    On Debian systems you probably can find a version of GPLv3 in
    /usr/share/common-licenses/
'''

from supybot.test import *


class KoranFinderTestCase(PluginTestCase):
    plugins = ('KoranFinder',)


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


pygq.py
"""
    Copyright 2018 Farooq Karimi Zadeh <farooghkarimizadeh@gmail.com>

"""

import requests

class PyGQ:
    SURAH_LAST = 114
    SURAH_FIRST = 1
    def __init__(self, url="http://api.globalquran.com/ayah/", token="", lg_codes={}):
        """url and token are API url and token. lg_codes is a dict which maps 
           2 letter language codes to their fullname. like:
           {"en": "en.sahih", "ar": "quran-simple"}
        """

        self.url = url
        self.token = token
        self.lg_codes = lg_codes

    def getAyah(self, surah, ayah, lang):
        """Returns a dict containing the verse itself, ayah and surah number
        and id. example:
            {
            "surah": 1,
            "ayah": 1,
            "id": 1,
            "verse": "In the name of Allah, the Entirely Merciful, the Especially Merciful."
            }
        """
        if len(lang) == 2:
            if lang in self.lg_codes:
                lang = self.lg_codes[lang]
            else:
                raise ValueError(lang+ " is not supported using 2letter codes")
        
        if (not self.SURAH_FIRST <= surah <= self.SURAH_LAST):
            raise ValueError("surah(chapter) must be between "
                             + str(self.SURAH_FIRST) + " and " +
                             str(self.SURAH_LAST))

        req_url = self.url + str(surah) + ":" + str(ayah) + "/" + lang
        ayah_json = requests.get(req_url, params = {'key': self.token}).json()
        while len(ayah_json) == 1:
            ayah_json = ayah_json[next(iter(ayah_json))]

        if ayah_json["surah"] != surah:
            raise ValueError("Invalid ayah number for this surah")
        
        return ayah_json

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Usage:", sys.argv[0], "surah:ayah [lang]")
        sys.exit(0)
    
    surah, ayah = sys.argv[1].split(":")
    surah = int(surah)
    ayah = int(ayah)
    lang = "en.sahih" if len(sys.argv) < 3 else sys.argv[2]

    Q = PyGQ()
    print(Q.getAyah(surah, ayah, lang)["verse"])
    


config.py
'''
    Copyright 2018 Farooq Karimi Zadeh <farooghkarimizadeh@gmail.com

    This file is part of KoranFinder.

    KoranFinder is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this software.  If not, see <http://www.gnu.org/licenses/>.

    On Debian systems you probably can find a version of GPLv3 in
    /usr/share/common-licenses/
'''

import supybot.conf as conf
import supybot.registry as registry
try:
    from supybot.i18n import PluginInternationalization
    _ = PluginInternationalization('KoranFinder')
except:
    # Placeholder that allows to run the plugin on a bot
    # without the i18n module
    _ = lambda x: x


def configure(advanced):
    # This will be called by supybot to configure this module.  advanced is
    # a bool that specifies whether the user identified themself as an advanced
    # user or not.  You should effect your configuration by manipulating the
    # registry as appropriate.
    from supybot.questions import expect, anything, something, yn
    KoranFinder = conf.registerPlugin('KoranFinder', True)
    if yn("""Split long verses?""", default=True):
        KoranFinder.splitMessages.setValue(True)
    else:
        KoranFinder.splitMessages.setValue(False)

KoranFinder = conf.registerPlugin('KoranFinder')
# This is where your configuration variables (if any) should go.  For example:
conf.registerGlobalValue(KoranFinder, 'splitMessages',
    registry.Boolean(True, _("""Set to split long verses.""")))
conf.registerGlobalValue(KoranFinder, 'splitMessagesAt',
    registry.PositiveInteger(350, _("""Set where to split long verses.""")))


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


__init__.py
'''
    Copyright 2017-2018 Farooq Karimi Zadeh <farooghkarimizadeh@gmail.com>

    This file is part of KoranFinder.

    KoranFinder is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this software.  If not, see <http://www.gnu.org/licenses/>.

    On Debian systems you probably can find a version of GPLv3 in
    /usr/share/common-licenses/
'''

"""
KoranFinder: Brings verses from Quran
"""

import supybot
import supybot.world as world

__version__ = "2018.5.27"

__author__ = supybot.Author("Farooq Karimi Zadeh", "FarooqKZ",
        "farooghkarimizadeh@gmail.com")

__contributors__ = { # Let's don't forget that dictionaries don't have order!
        supybot.Author("##islam channel member on Freenode"),
        supybot.Author("Antti Siponen", "lantti"),
        supybot.Author("Usama Akkad", "damascene", "uahello@gmail.com"),
        supybot.Author("Safa Alfulaij", "SafaAlfulaij",
            "safaalfulaij@hotmail.com"),
        supybot.Author("Farooq Karimi Zadeh", "FarooqKZ",
            "farooghkarimizadeh@gmail.com")
        } 

__url__ = 'https://github.com/farooqkz/KoranFinder'

from . import config
from . import plugin
from imp import reload
# In case we're being reloaded.
reload(config)
reload(plugin)
# Add more reloads here if you add third-party modules and want them to be
# reloaded when this plugin is reloaded.  Don't forget to import them as well!

if world.testing:
    from . import test

Class = plugin.Class
configure = config.configure


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


plugin.py
'''
    Copyright 2017-2018 Farooq Karimi Zadeh <farooghkarimizadeh@gmail.com>

    This file is part of KoranFinder.

    KoranFinder is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this software.  If not, see <http://www.gnu.org/licenses/>.

    On Debian systems you probably can find a version of GPLv3 in
    /usr/share/common-licenses/
'''

import supybot.utils as utils
from supybot.commands import *
import supybot.plugins as plugins
import supybot.ircutils as ircutils
import supybot.callbacks as callbacks
from . import pygq
try:
    from supybot.i18n import PluginInternationalization
    _ = PluginInternationalization('KoranFinder')
except ImportError:
    # Placeholder that allows to run the plugin on a bot
    # without the i18n module
    _ = lambda x: x

# Set languages and translations the bot will accept
quranID = {"ar" : "quran-simple", "en" : "en.sahih", "tr" : "tr.yazir", "fa" : "fa.fooladvand"}

API_TOKEN = ""

class utilities():
    def arg2list(ayat): # TODO: better name for this function
        MAX_AYAT = 6
        ayat = ayat.split(",")
        ayat_l = []
        for aya in ayat:
            if aya.isdigit():
                ayat_l += [int(aya)]
            else:
                t = aya.split('-') 
                start, end = int(t[0]), int(t[1])
                ayat_l += list(range(start, end + 1))

        if len(ayat_l) > MAX_AYAT:
            raise ValueError("Sorry, you can get just " + str(MAX_AYAT) +
                    " ayat each time calling me.")
        else:
            return ayat_l

class KoranFinder(callbacks.Plugin):
    """This plugin gets verse and ayah number and sends you the ayah using a web API."""
    threaded = True

    def __init__(self, irc):
         self.__parent = super(KoranFinder, self)
         self.__parent.__init__(irc)
         # TODO: let user config cache_size
         self.PyGQ = pygq.PyGQ(token = API_TOKEN, lg_codes=quranID,cache_size=100)

    def koran(self, irc, msg, args, surah, ayat, lang):
        """<surah> <ayah/ayat> <lang>

        returns ayah number <ayah> of surah number <surah> in <lang> language or translation or tafsir. for more information visit: https://git.io/vwMz9
        """
        
        try:
            list_of_ayat = utilities.arg2list(ayat)
        except ValueError as e:
            irc.error(str(e))
            return

        try:
            ayat_list = []
            for ayah in list_of_ayat:
                verse_json = self.PyGQ.getAyah(surah, ayah, lang)
                ayat_list.append(str(verse_json["surah"]) + ":" +
                                str(verse_json["ayah"]) + ", " +
                                str(verse_json["verse"]))
        except ValueError as e:
            irc.error(str(e))
            return
        except (KeyError, TypeError) as e: #TypeError incase requesting a audio version. The json would contain a list so qdata would raise a TypeError
            irc.error("Wrong translation code or broken API.") 
            return
        for ayah in ayat_list:

            if self.registryValue('splitMessages'):
                ircMsgBytes = ayah.encode('utf-8')
                while len(ircMsgBytes) > 350:
                    splitPoint = ircMsgBytes[0:351].rfind(' '.encode('utf-8'))
                    irc.reply(ircMsgBytes[0:splitPoint].decode('utf-8').strip())
                    ircMsgBytes = ircMsgBytes[splitPoint:]
                irc.reply(ircMsgBytes.decode('utf-8').strip())
            else:
                irc.reply(ayah)

    koran = wrap(koran, ["int", "something", optional("something", "en")])

Class = KoranFinder
# vim:set shiftwidth=4 softtabstop=4 expandtab textwidth=79:
